contracts/CommunityTokenReward.sol

```diff
@@ -0,0 +1,170 @@
+
+  constructor(address tokenAddress, uint256 maxRewardPerPeriod, uint256 maxRewardPerUser, address communityTokenRewardFactoryAddress, uint32 communityId, address peeranhaUserContractAddress) {
+    // if address = 0?
+    communityTokenContainer.info.tokenAddress = tokenAddress;
+    communityTokenContainer.info.name = IERC20Metadata(communityTokenContainer.info.tokenAddress).name();
+    communityTokenContainer.info.symbol = IERC20Metadata(communityTokenContainer.info.tokenAddress).symbol();
+    communityTokenContainer.info.maxRewardPerPeriod = maxRewardPerPeriod;
+    communityTokenContainer.info.maxRewardPerUser = maxRewardPerUser;
+    communityTokenContainer.info.createTime = CommonLib.getTimestamp();
+    communityTokenContainer.info.communityTokenRewardFactoryAddress = communityTokenRewardFactoryAddress;
+    communityTokenContainer.info.communityId = communityId;
+    communityTokenContainer.peeranhaUser = IPeeranhaUser(peeranhaUserContractAddress);
+  }
+
+  // This is to support Native meta transactions
+  // never use msg.sender directly, use _msgSender() instead
+  function _msgSender()
+      internal
+      override(Context, NativeMetaTransactionNonUpgrade)
+      view
+      returns (address sender)
+  {
+    return NativeMetaTransactionNonUpgrade._msgSender();
+  }
+
+  function dispatcherCheck(address userAddress) internal {
+    if (userAddress != _msgSender()) {
+      communityTokenContainer.peeranhaUser.onlyDispatcher(_msgSender());
+    }
+  }
+
+  function updateCommunityRewardSettings(address userAddress, uint256 maxRewardPerPeriod, uint256 maxRewardPerUser) external override {
+    dispatcherCheck(userAddress); 
+    communityTokenContainer.peeranhaUser.checkHasRole(_msgSender(), UserLib.ActionRole.CommunityAdmin, communityTokenContainer.info.communityId); // todo tests
+
+    communityTokenContainer.info.maxRewardPerPeriod = maxRewardPerPeriod;
+    communityTokenContainer.info.maxRewardPerUser = maxRewardPerUser;
+    emit CommunityRewardSettingsUpdated();
+  }
+
+  function getAvailableRewardsBalance() public view override returns(uint256) {   // public?
+    uint256 balance = IERC20(communityTokenContainer.info.tokenAddress).balanceOf(address(this));
+    if (balance <= communityTokenContainer.info.reservedTokens) {
+      return 0;
+    }
+
+    return balance - communityTokenContainer.info.reservedTokens;
+  }
+
+  // get pool
+  function getPeriodRewardPatams(uint16 period) public view returns(uint256 maxTotalTokenPool, uint256 totalTokenPool, uint256 maxTokensPerUser, uint256 balance) {
+    RewardPeriodParams storage rewardPeriodParams = communityTokenContainer.rewardPeriodParams[period];
+    return (rewardPeriodParams.maxTotalTokenPool, rewardPeriodParams.totalTokenPool, rewardPeriodParams.maxRewardPerUser, rewardPeriodParams.availableBalance);
+  }
+
+  // set pool
+  function startNewPeriod(RewardLib.PeriodRewardShares memory periodRewardShares, uint16 period) external override {
```

Comment:
This is not clear that `periodRewardShares` is not for  `period`

`newPeriod`, `claimPeriodRewardShares`

Also do we need to pass RewardLib.PeriodRewardShares? can pass just number of active users in the period?